vllm/entrypoints/openai/serving_embedding.py

```diff
@@ -0,0 +1,150 @@
+        data=data,
+        usage=usage,
+    )
+
+
+def merge_async_iterators(*iterators):
+    """Merge multiple asynchronous iterators into a single iterator.
+
+    This method handle the case where some iterators finish before others.
+    When it yields, it yields a tuple (i, item) where i is the index of the
+    iterator that yields the item.
+    """
+    queue = asyncio.Queue()
+
+    finished = [False] * len(iterators)
+
+    async def producer(i, iterator):
+        async for item in iterator:
+            await queue.put((i, item))
+        finished[i] = True
+
+    _tasks = [
+        asyncio.create_task(producer(i, iterator))
+        for i, iterator in enumerate(iterators)
+    ]
+
+    async def consumer():
+        while not all(finished) or not queue.empty():
+            item = await queue.get()
+            yield item
+        await asyncio.gather(*_tasks)
+
+    return consumer()
+
+
+class OpenAIServingEmbedding(OpenAIServing):
+
+    def __init__(self, engine: AsyncLLMEngine, served_model: str):
+        super().__init__(engine=engine,
+                         served_model=served_model,
+                         lora_modules=None)
+
+    async def create_embedding(self, request: EmbeddingRequest,
+                               raw_request: Request):
+        """Completion API similar to OpenAI's API.
+
+        See https://platform.openai.com/docs/api-reference/embeddings/create
+        for the API specification. This API mimics the OpenAI Embedding API.
+        """
+        error_check_ret = await self._check_model(request)
+        if error_check_ret is not None:
+            return error_check_ret
+
+        # Return error for unsupported features.
+        if request.encoding_format == "base64":
+            return self.create_error_response(
+                "base64 encoding is not currently supported")
+        if request.dimensions is not None:
+            return self.create_error_response(
+                "dimensions is currently not supported")
+
+        model_name = request.model
+        request_id = f"cmpl-{random_uuid()}"
+        created_time = int(time.monotonic())
+
+        # Schedule the request and get the result generator.
+        generators = []
+        try:
+            prompt_is_tokens, prompts = parse_prompt_format(request.input)
+
+            for i, prompt in enumerate(prompts):
+                if prompt_is_tokens:
+                    input_ids = self._validate_prompt_and_tokenize(
+                        request, prompt_ids=prompt)
+                else:
+                    input_ids = self._validate_prompt_and_tokenize(
+                        request, prompt=prompt)
+
+                generators.append(
+                    self.engine.generate(prompt,
+                                         SamplingParams(),
+                                         f"{request_id}-{i}",
+                                         prompt_token_ids=input_ids))
+        except ValueError as e:
```

Comment:
If this exception is thrown after some generators are created, shouldn't we cancel them? Or perhaps first prep the input_ids for all sequences, and then just append the generators (which shouldn't throw).

Reply:
Same as above. The logic here is the same as [serving_completion](https://github.com/vllm-project/vllm/blob/main/vllm/entrypoints/openai/serving_completion.py#L151-L159). 
The requests with exceptions thrown here are not thrown by appending the generators, right?

Reply:
Well, by looking at the code the append line appends an un-awaited promise. The promise, if not awaited is not even executed (the code that fulfills it, that is), so it shouldn't be able to throw anything. The other lines however throw stuff, if there is an issue with the input.

Thinking about this -- if an exception is throw the generators are not awaited, so they shouldn't actually deploy any work to the engine, so I guess that part of the code is fine.